lib/temporal/worker/reactor.rb

```diff
@@ -0,0 +1,112 @@
+require 'fiber'
+require 'forwardable'
+require 'set'
+require 'singleton'
+require 'temporal/worker/reactor/promise'
+require 'temporal/worker/reactor/command'
+
+module Temporal
+  class Worker
+    class Reactor
+      include Singleton
+      extend SingleForwardable
+
+      def_delegators :instance, :execute, :attach
+
+      def initialize
+        @queue = Queue.new
+        @fibers = Set.new
+        @thread = nil
+        @running = false
+        @mutex = Mutex.new
+      end
+
+      def execute(&block)
+        mutex.synchronize do
+          unless running?
+            @thread = Thread.new { run_loop }
+            @running = true
+          end
+        end
+
+        enqueue { block.call(self) }
+      end
+
+      def async(&block)
+        promise = Promise.new
+
+        enqueue do
+          fiber = Fiber.current
+
+          resolver = lambda do |value = nil|
+            resume(fiber, value)
+          end
+
+          block.call(resolver)
+
+          value = Fiber.yield
+          promise.set(value)
+        end
+
+        promise
+      end
+
+      def await(&block)
+        promise = async(&block)
+        promise.result
+      end
+
+      def attach
+        thread&.join
+      end
+
+      private
+
+      attr_reader :queue, :thread, :fibers, :mutex
+
+      def running?
+        @running
+      end
+
+      def run_loop
+        loop do
+          command = queue.pop
+
+          handle(command)
+
+          fibers.delete_if { |f| !f.alive? }
+
+          mutex.synchronize do
+            if queue.empty? && fibers.empty?
```

Comment:
What if something is enqueued on the unsynchronized queue after this conditional?

I don't think you need to break here. One thread being held by the system until shutdown that hangs on `queue.pop` is probably fine (granted a fiber-capable queue would be ideal, though I haven't spec'd out in my head what that might look like).

Reply:
Ah, good point, missed this.

And yes, we can probably keep the loop running forever and not worry about stopping it